dockerd gpu support

flx42 opened the pull request:
API discussion: https://github.com/moby/moby/pull/37434
CLI discussion: https://github.com/docker/cli/issues/1200

This is a simple prototype for GPU support in `dockerd`. It uses a simple in-tree plugin mechanism similar to the graphdriver plugins in `daemon/graphdriver/...`, but stateless.

## Plugin API
Since plugins have access to the full OCI spec, I expect plugins to do one (or multiple) of the following:
* Add mounts/devices/environments variables to the OCI spec.
* Add hooks to the OCI spec.
* Set OCI annotations that will be picked up later by a component closer to the container runtime implementation (containerd, runc, kata?). This is for complex cases that are largely outside of the OCI spec.

**Is this acceptable? Or do you want to restrict which part of the spec the plugins can edit?**

## Discussion
The demo implementation for the `nvidia` plugin in this PR relies on the existing `nvidia-container-runtime-hook` binary since I'm not  sure yet how you think we should leverage the recently added GPU support in containerd.
1. Should it be part of this GPU plugin API as another handler?
```
import "github.com/containerd/containerd/oci"
import "github.com/opencontainers/runtime-spec/specs-go"
func ContainerdHandler(spec specs.Spec, opts oci.SpecOpts)
```
This feel a bit awkward since we're not sure what other implementations could look like.
If we do this, we technically could get rid of OCI annotations since everything is handled in `dockerd`, but I think it's good to keep a trace of what happened in `config.json`, even if it's plain annotations.

2. Should we try to standardize on the format of the OCI annotations for GPUs and thus add the code for handling these annotations in `containerd` directly?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "dockerd-gpu-support" git@github.com:flx42/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353798760
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

flx42 commented:
@cpuguy83 @crosbymichael

crosbymichael commented on libcontainerd/client_daemon.go:
```diff
@@ -213,6 +215,29 @@ func (c *client) Restore(ctx context.Context, id string, attachStdio StdioCallba
 
+// The problem with this approach is the split logic between the plugin-like code in daemon/gpu/nvidia
+// and the containerd client code which needs to embed the GPU API calls directly.
+func setGPUOptions(ociOpts []oci.SpecOpts, ociSpec *specs.Spec) []oci.SpecOpts {
```
Just call this code in the handler.  You can call the spec opts like any other function.  You can just make a simple combinator for this.

in your gpu handler dont use annotations and just process the fields like:

```go
func applyOpts(s *specs.Spec, opts ...SpecOpts) {
    for _, o := range opts {
         o(ctx, nil, nil, s)
    }
}
```

crosbymichael replied:
or you just build your gpu opts and call `nvidia.WithGPUs(nvidiaOpts)(ctx, nil, nil, spec)`

flx42 committed:
[HACK] revendor

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

flx42 committed:
api: add a GPU resource type

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

flx42 committed:
Do not override existing OCI hooks

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

flx42 committed:
Add simple in-tree plugin mechanism for GPU support

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

flx42 committed:
Export the name of the containerd binary

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

flx42 commented:
This new feature is now behind the `experimental` flag of dockerd, as requested.
It's in a functional state today and the approach is straightforward, please help review the design/code. I just need to add a few unit tests.

flx42 commented:
Removed the WIP tag from the title

cpuguy83 commented on daemon/gpu/gpu.go:
```diff
@@ -0,0 +1,40 @@
+type SpecHandler func(spec *specs.Spec, set container.GPUSet) error
+
+func init() {
+	vendors = make(map[string]SpecHandler)
```
Don't think you need to do this in a `func init()`

flx42 replied:
Need to double-check the order of initialization. I did the same as:
https://github.com/moby/moby/blob/master/daemon/graphdriver/driver.go#L148-L150

cpuguy83 commented on daemon/oci_linux.go:
```diff
@@ -76,10 +77,28 @@ func setResources(s *specs.Spec, r containertypes.Resources) error {
 
+func setGPUResources(daemon *Daemon, s *specs.Spec, r []containertypes.GPUSet) error {
+	if len(r) > 0 && !daemon.HasExperimental() {
+		return fmt.Errorf("GPUs are only supported in experimental mode")
```
Please use an error from the `errdefs` package.

cpuguy83 commented on daemon/oci_linux.go:
```diff
@@ -76,10 +77,28 @@ func setResources(s *specs.Spec, r containertypes.Resources) error {
 	return nil
 }
 
+func setGPUResources(daemon *Daemon, s *specs.Spec, r []containertypes.GPUSet) error {
```
Seems like the experimental check should be higher up in the stack, and greatly simplify this function.

flx42 replied:
Ok, and do you prefer passing a boolean to `setResources` instead of the `daemon` object?

cpuguy83 replied:
No, I would just not call `setGPUResources` if the daemon is not experimental.

flx42 replied:
Sure, but I'm talking about the function one step up: `setResources`, this function needs to be called in all cases.

cpuguy83 replied:
Ah... a little gross but... maybe we can do `setResources` and `setResourcesExperimental`?

flx42 replied:
seems reasonable.

cpuguy83 commented on libcontainerd/client_daemon.go:
```diff
@@ -563,7 +563,8 @@ func (c *client) CreateCheckpoint(ctx context.Context, containerID, checkpointDi
 	if exit {
-		opts = append(opts, containerd.WithExit)
+		// Revendoring hack: please disregard.
+		//opts = append(opts, containerd.WithExit)
```
🤔

flx42 replied:
As it says on the tin, sorry for the revendoring noise :)

cpuguy83 commented on daemon/gpu/nvidia/nvidia.go:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	var opts []nvidia.Opts
+	opts = append(opts, nvidia.WithLookupOCIHookPath(libcontainerd.BinaryName))
```
Can you explain this?

flx42 replied:
The default behavior from Michael's `nvidia` plugin is to use `containerd` as the binary. The hook is executed as:
```
containerd oci-hook -- nvidia-container-cli ... [args] ...
```
Maybe it should have been a separate binary, but that's the status right now.

flx42 replied:
So in this case we need `docker-containerd`, not `containerd`.

cpuguy83 commented on daemon/gpu/nvidia/nvidia.go:
```diff
@@ -0,0 +1,112 @@
+
+func handler(spec *specs.Spec, set container.GPUSet) error {
+	if set.Vendor != "nvidia" {
+		return fmt.Errorf("vendor mismatch: received %s, expected nvidia", set.Vendor)
```
errdefs

cpuguy83 commented on daemon/gpu/nvidia/nvidia.go:
```diff
@@ -0,0 +1,112 @@
+	opts = setDriverCapabilities(opts, spec, set)
+	opts = setCUDAVersion(opts, spec, set)
+
+	return nvidia.WithGPUs(opts...)(nil, nil, nil, spec)
```
context should never be passed as a nil

cpuguy83 commented on daemon/oci_linux.go:
```diff
@@ -712,7 +731,7 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
 	s.Linux.CgroupsPath = cgroupsPath
 
-	if err := setResources(&s, c.HostConfig.Resources); err != nil {
+	if err := setResources(daemon, &s, c.HostConfig.Resources); err != nil {
```
🤔

flx42 replied:
This is what I was mentioning above, do you want to pass a boolean here instead of the daemon?

flx42 commented:
Updated my PR to make the changes requested by @cpuguy83. I dropped the idea of `setResourcesExperimental` since it was too ugly, I pass a boolean flag to `setResources`.

The vendoring is still hacky since this is not the right place to do it I believe, and my understanding is that we would want to wait for a new minor/patch version of containerd anyway.

flx42 committed:
Add support for NVIDIA GPUs in dockerd

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

cpuguy83 commented on daemon/oci_linux.go:
```diff
@@ -76,10 +78,30 @@ func setResources(s *specs.Spec, r containertypes.Resources) error {
 	}
 
+	if len(r.GPUs) > 0 {
+		if !experimental {
```
Would it make sense to push this experimental check up the stack so we don't have to deal with it here?
Also, do you really feel this should be experimental? Are you worried the API is not sufficient?

flx42 replied:
> Would it make sense to push this experimental check up the stack so we don't have to deal with it here?

I initially considered what you said previously, that is to say to create a function `setResourcesExperimental` and put it here after `setResources`:
https://github.com/moby/moby/blob/658675e7d6fff1e4ef7c153398c25fbbe52a5080/daemon/oci_linux.go#L737-L739
Then I had two options:
```golang
if daemon.HasExperimental() {
        if err := setResourcesExperimental(...) {
                return nil, fmt.Errorf(...)
        }
}
```
That's not great since we lose the ability to warn/error if GPUs are requested but the experimental flag is not set.

The other option would have been:
```
if err := setResourcesExperimental(...,  daemon.HasExperimental()) {
        return nil, fmt.Errorf(...)
}
```
But it's bizarre to have a function `setResourcesExperimental` and pass a boolean flag for this function to really do something.

Let me know if you see another option.

flx42 commented:
>  Also, do you really feel this should be experimental? Are you worried the API is not sufficient?

Moving this conversation to the main thread since it's a more fundamental question.
@crosbymichael asked me to put this PR behind this experimental flag, the rationale was that the multiple APIs might be unstable:
1) The CLI API, i.e. `--gpus`. This is clearly something we don't want to break if it's not behind an experimental flag. I think we mostly agreed on the format, except maybe for how it would map to Compose.
2) The `docker` API. Same, we don't want to break this API if it's not experimental. But I think what we have today is simple and extensible, so that should be fine.
3) The internal plugin API. The input is the OCI runtime spec, the output is a modified OCI runtime spec. This is not very likely to change, and it's actually not a big deal if it does since it's not a public API.

But for all of the above, with my implementation being the sole client/consumer, I would understand gating that behind an `experimental` flag. I would be happy to change it back.

crosbymichael commented:
@thaJeztah how does vendoring work when there are client and server changes?

thaJeztah commented:
@crosbymichael you mean if changes are needed both in the CLI and on the daemon side?

flx42 commented:
I guess that's what he meant. My guess is that the ordering needs to be:
1. Release of a new tagged version of `containerd` (e.g. `1.1.3` or `1.2.0`)
2. Revendoring of `containerd` in `moby` to get the latest GPU patches from @crosbymichael (bundled in this PR in a very hacky way)
3. Merge of the API change: https://github.com/moby/moby/pull/37434 (also bundled in this PR)
4. Merge of the `dockerd` implementation for this new API (this PR)
5. Merge of the CLI changes: https://github.com/docker/cli/issues/1200

Is that correct?

flx42 commented:
Ping @thaJeztah 

@crosbymichael do you confirm we should gate this feature behind the `experimental` flag?

flx42 commented:
@thaJeztah @crosbymichael since I imagine we're now close to the `18.09` release and the built containerd is still `1.1.2`, are we targeting this feature for docker CE `19.03`?

olljanat commented:
@flx42 this one needs to be rebased

olljanat commented:
Derek add label: status/failing-ci

tiborvass commented:
Closed by #38828
